Package API for torch.compile

zhxchen17 opened the pull request:
Package API for torch.compile

Following up PR https://github.com/pytorch/pytorch/pull/145381, we implement
a new API for compiling models using the cpp wrapper, and save/load
compiled artifacts to disk.

Package is now designed to be a per-torch.compile() object living with
the compilation context. Each time a recompilation happens, it will collect
the compiled artifacts into a lookup table. When a new set of inputs is
passed to the compiled callable, before we enter the dynamo cache, we will
perform a lookup in compile package first, and match by the serialized guards.

API names are tentative but the workflow roughly looks like the following:

```
def f(...): ...

compiled_f = torch.compile(f, package="my_dir/my_model")

compiled_f(*args)

compiled_f.save_package(prefix="/dir1")

...

compiled_f.load_package(prefix="/dir2")
```


Fixes #ISSUE_NUMBER


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

jansel commented on torch/__init__.py:
```diff
@@ -2551,13 +2588,24 @@ def fn(model: _Callable[_InputT, _RetT]) -> _Callable[_InputT, _RetT]:
+        from torch._dynamo.sticky_cache import _StickyCache
+
+        sticky_cache = _StickyCache(sticky_cache)  # type: ignore[assignment]
+        if dynamic is not False:
+            sticky_cache.unimplemented("dynamic shape")  # type: ignore[attr-defined]
```
Dynamic is not the only way to get dynamic shapes.   Can we support dynamic shapes by using dynamo guards?

zhxchen17 replied:
I can add dynamo guards in this PR since it seems to easier to just support it rather than patching `unimplemented()` around.

jansel commented on torch/_dynamo/sticky_cache.py:
```diff
@@ -0,0 +1,271 @@
+      - AOTI compiled code and kernels for forward and backward graph.
+      - AOTAutograd wrappers for things like input mutation.
+      - Dynamo bytecode for mapping Python inputs/outputs.
+      - Dynamo guards for cache keys.
```
Where are the dynamo guards saved?

jansel commented on torch/_dynamo/sticky_cache.py:
```diff
@@ -0,0 +1,271 @@
+        assert self.dynamo_code is None
+        self.dynamo_code = dynamo_code
+
+    def match_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> bool:
+        # TODO use dynamo guards when it's ready
+        assert self.inputs_metadata is not None
+        for a, b in zip(self.inputs_metadata, _extract_inputs_metadata(args, kwargs)):
+            if a != b:
+                return False
+        return True
```
Can we reuse dynamo guards to make dynamic shapes work?  cc @anijain2305

jansel commented on torch/_dynamo/sticky_cache.py:
```diff
@@ -0,0 +1,271 @@
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        assert self.dynamo_code is not None
+        name = next(
+            n for n in self.dynamo_code.co_names if n.startswith("__compiled_fn")
+        )
+        return types.FunctionType(self.dynamo_code, globals={name: self.aoti})(
+            *args, **kwargs
+        )
```
This seems wrong.

Dynamo code might depend on other globals in addition to `__compiled_fn0`, both in user code and in generated code.  I think we need to examine the co_names of the code object.

This will also be incorrect if you compile two functions in the same file since the second will be `__compiled_fn1`.

zhxchen17 replied:
I guess an example would be helpful and will be happy to add it to test.

jansel replied:
You can fullgraph compile two different functions in the same file (or two different shapes for the same functions).

For other globals, put a tensor in global scope and read from it.

zhxchen17 commented on torch/__init__.py:
```diff
@@ -2509,6 +2518,34 @@ def compile(
+        with fullgraph=True. When sticky_cache is not None, there will be additional methods on the compiled object
+        to control serialization of compiled code:
+
+        - .save_sticky_cache(prefix): Save compiled code to directory prefixed by `prefix`. `prefix` will be
+                                      empty if not given.
+        - .load_sticky_cache(prefix): Load compiled code from directory prefixed by `prefix`. `prefix` will be
+                                      empty if not given.
```
I guess it will be helpful for the case where we save it on one directory of one machine and load it from another path on a different machine, when we are sharing the same torch.compile() configs from the saving and loading side.

zhxchen17 replied:
but I don't think we must include this in version 1. will remove.

zhxchen17 commented on torch/__init__.py:
````diff
@@ -2509,6 +2518,34 @@ def compile(
+        Example workflow:
+        ```
+        def fn(…): …
+        compiled_fn = torch.compile(fn, fullgraph=True, sticky_cache="dir/model”)
````
No strong preference here, I can make the default auto save/load and opt out with a compiler config.

zhxchen17 commented on test/test_sticky_cache.py:
```diff
@@ -0,0 +1,328 @@
+
+        args = torch.randn(3, requires_grad=True), torch.randn(3, requires_grad=True)
+        f = torch.compile(f, fullgraph=True, dynamic=False, sticky_cache=self.path())
+        with self.assertRaisesRegex(
+            NotImplementedError,
+            "backward",
+        ):
+            f(*args).backward()
```
Can we leave backward unimplemented for this PR? For inference case I think we can already make good progress even with forward() only. Supporting backward will bloat the scope of this PR a bit and might block the demos for torchnative.

zhxchen17 commented on torch/__init__.py:
````diff
@@ -2509,6 +2518,34 @@ def compile(
+        # No recompile if and only if the loaded cache was compiled under inputs with compatible tensor shape, device, etc.
+        compiled_fn(*example_inputs)
+        ```
+        [NOTE] Loading a precompiled sticky cache has some soundness implication due to the fact that
+        only a subset of the original guards are checked. For example, if the compiled model relies on a global
+        variable to be in specific state, loaded compile may simply ignore them even if they have been changed at
+        loading time. Nonetheless, guards on model inputs will guaranteed to be checked.
````
Do you mean we should put this as a warning for loading?

oulgen commented:
@zhxchen17 @jansel  I have a bit of a meta point, can we NOT call the public API a cache? We have term overload where people tend to come back saying "cache is not working", but there's always ambiguity, is it the the dynamo recompilation cache? Is it inductor/aotautograd cache? and now is it the sticky cache?

I feel like there will be a good value in clear disambiguation.

oulgen commented:
> > @zhxchen17 @jansel I have a bit of a meta point, can we NOT call the public API a cache? We have term overload where people tend to come back saying "cache is not working", but there's always ambiguity, is it the the dynamo recompilation cache? Is it inductor/aotautograd cache? and now is it the sticky cache?
> > we already have `torch.compiler.load_cache_artifacts` and `torch.compiler.save_cache_artifacts` which makes this further confusing.
> > I feel like there will be a good value in clear disambiguation.
> 
> @oulgen I got your point. Do you like the naming "persistent_artifacts" better than sticky_cache? I can switch name to avoid the confusion here.

yes, that sounds a lot better, thanks

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -737,6 +738,13 @@ def _dup_fake_script_obj(fake_flat_args):
 
+        if aot_config.sticky_cache is not None:
+            if any(
+                info.mutation_type != MutationType.NOT_MUTATED
```
let me know if you want more review on the AOTAutograd bits (we might want to think more about how this interacts with the existing AOT warm cache today).

One comment is that input mutations are probably fine for sticky cache, as long as the input mutation is captured inside of the graph (the "bad" case is if the input mutation is forced to run outside of the graph, in an opaque AOTAutograd epilogue). Almost all input mutations at this point are captured in-graph today

jamesjwu commented on torch/_dynamo/compile_package.py:
```diff
@@ -0,0 +1,756 @@
+
+class _GraphCompile:
+    """
+    Stores the compiled artifacts per compiled FX graph. This includes:
```
This is not exactly accurate, is it? It stores the compiled artifacts for N compiled FXGraphs, not just one, unless you mean the input function, which itself is not an FXGraph?

jamesjwu commented on torch/_dynamo/compile_package.py:
```diff
@@ -0,0 +1,756 @@
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self.forward_aoti: Optional[CompiledAOTI] = None
```
In my head, what would make most sense is that an AOTAutogradCacheEntry contains a forward/backward_aoti. I know this is a prototype so you probably didn't want to change AOTAutogradCacheEntry, but I think what you'd actually want is:
- Refactor AOTAutogradCacheEntry into an interface witha  forward/backward OutputCode + the aot autograd wrappers. You could make it generic on the OutputCode type, i.e. AOTAutogradCacheEntry[CompiledAOTI] vs. AOTAutogradCacheEntry[CompiledFXGraph], etc. 
 
- Split AOTAutogradCacheEntry into the cache path, where there's a python wrapper + FXGraphCache keys, vs. the package API, which uses AOTI artifacts

Then here, you would just store self.aot_autograd_output : AOTAutogradCacheEntry[CompiledAOTI].

jamesjwu commented on torch/_dynamo/compile_package.py:
```diff
@@ -0,0 +1,756 @@
+        aot_config = pickle.load(f)
+
+        def _(*args, **kwargs):
+            raise RuntimeError("NYI")
```
You're never gonna be able to serialize these functions, so they're not NYI, right? They're actually full forward/backward compilers(often inductor libraries). Instead I think you want to ignore these callables (or assume that they are always the same)

jamesjwu commented on torch/_dynamo/compile_package.py:
```diff
@@ -0,0 +1,756 @@
+#      second half, and serialier should read data from the return
+#      value of the first half.
+#      For prototyping we just leave this as a separate class.
+class GuardSerializer:
```
What do we think about moving the guard serde into its own PR? That way you could test it separately, and have it be generic, rather than specific to the package API

oulgen commented:
@zhxchen17 I think it might be good to discuss the `@torch.compile(package="/package/path")`, I suspect having the package name on the compile API is OK for most OSS jobs, but I suspect will make it difficult to apply it on mast jobs, so it might be worthwhile moving this to load/save_package calls

jamesjwu commented on torch/_dynamo/compile_package.py:
```diff
@@ -0,0 +1,756 @@
+                value_len, get_verbose_code_parts(code, guard)
+            )
+
+    def HASATTR(self, guard, metadata):
```
what's the protocol for making sure people don't add new guards without adding this?

Would it be simpler not to have each Guard have a ser/deser method? That way, when people add new Guard types, they would have to implement serde directly, instead of finding a separate class for them.

jamesjwu reviewed (commented):
Really cool! Some initial questions about code organization, at least in the higher level sections I understand haha

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+        load_paths = glob.glob(os.path.join(load_dir, "*.dynamo_code"))
+        self.assertEqual(len(load_paths), 0)
+
+        compiled_fn.save_package()
```
Why not take the path here?  Seems a bit weird to have the path be an arg to torch.compile (which won't do anything unless you call save).

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+
+        precompiles = []
+        for i in range(len(load_paths)):
+            precompile = torch._dynamo.compile_package._load_precompile(load_dir, i)
```
How is this different that compiled_fn.load_package?

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+        with self.assertRaisesRegex(
+            RuntimeError, "Compile package is only supported .*fullgraph=True"
+        ):
+            torch.compile(f, dynamic=False, package=self.path())
```
I'm a bit confused about how `package=` changes the semantics of torch.compile?  Ideally there would be no semantic differences (in which case the path could just be an arg to save/load_package).

zhxchen17 replied:
as discussed yesterday, we there'll always be corner cases compile package doesn't work (e.g. unpicklable local classes), so we need something to make torch.compile() start to build the package only when asked to.

That being said, we can also provide a global boolean flag instead. (Might be easier for batch testing anyways)

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+            return a
+
+        with self.assertRaisesRegex(
+            RuntimeError, "Compile package is only supported .*fullgraph=True"
```
I want to confirm you plan to remove the limitation, since a lot of key users of this have graph breaks.

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+
+        with self.assertRaisesRegex(
+            NotImplementedError,
+            "dynamic shape",
```
This is planned correct?

zhxchen17 replied:
yep

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+        f = torch.compile(f, fullgraph=True, dynamic=False, package=self.path())
+        with self.assertRaisesRegex(
+            NotImplementedError,
+            "backward",
```
This is planned correct?

zhxchen17 replied:
yep, planned

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+        from torch._dynamo.exc import InternalTorchDynamoError
+
+        with self.assertRaisesRegex(
+            InternalTorchDynamoError,
```
This seems like the wrong exception type.

jansel commented on test/test_compile_package.py:
```diff
@@ -0,0 +1,374 @@
+        self._test_load(f, args3, expected3, self.path())
+
+        # self._test_save(f, (torch.randn(4, 2), torch.randn(4, 3)), None, self.path(), None)
+
```
Can we add a test for trying to load a saved package generated for function `foo` on function `bar`.
1) If the name/signature don't match, we should throw an error on load
2) We should also recursively hash the source code and assert the source codes match.  (With a flag to disable this check)

jansel commented on torch/__init__.py:
```diff
@@ -2292,11 +2292,15 @@ def compiled_with_cxx11_abi() -> builtins.bool:
         self.dynamic = dynamic
+        self.package = package
+        if self.package:
+            self.apply_options({"cpp_wrapper": True})
+            self.apply_options({"aot_inductor.package": True})
```
IMO the initial version of this should use the Python wrapper code.  I think this will be required (at least in the short term) to support:
1) Training
2) Dynamic shapes
3) Graph breaks

I think the use case of single-graph, static-shape, inference is already well covered by export+AOTI and doesn't really need a new API.

jamesjwu replied:
I think if we refactor the package code with AOTAutogradCacheEntry properly, we can combine these designs nicely into a single framework:
- Caching creates an AOTAutogradCacheEntry with Python wrapper cache entries 
- Precompile creates an AOTAutogradCacheEntry with either CPP wrapper or python wrapper cache entries, + a set of serialized dynamo guards. 

The nice thing about this is, if we do it right, we also get regular caching support for cpp wrapper, which we actually don't have today.

zhxchen17 replied:
@jansel I'm all for training as well. Just disagree with this

> inference is already well covered by export+AOTI and doesn't really need a new API.

inference is not well covered by export+AOTI since the UX is still broken for compiling partial graphs (wrote a post about this https://fb.workplace.com/groups/257735836456307/permalink/802058152024070/)

I'd like to still take inference into account, and as @jamesjwu mentioned, we can leverage caching to support both cases.

jansel replied:
I was talking about "single-graph, static-shape, inference" (all three at the same time) being well covered.

zhxchen17 committed:
Package API for torch.compile

Following up PR https://github.com/pytorch/pytorch/pull/145381, we implement
a new API for compiling models using the cpp wrapper, and save/load
compiled artifacts to disk.

Package is now designed to be a per-torch.compile() object living with
the compilation context. Each time a recompilation happens, it will collect
the compiled artifacts into a lookup table. When a new set of inputs is
passed to the compiled callable, before we enter the dynamo cache, we will
perform a lookup in compile package first, and match by the serialized guards.

API names are tentative but the workflow roughly looks like the following:

```
def f(...): ...

compiled_f = torch.compile(f, package="my_dir/my_model")

compiled_f(*args)

compiled_f.save_package(prefix="/dir1")

...

compiled_f.load_package(prefix="/dir2")
```

zhxchen17 commented:
> Is this still alive?

@ezyang Moving to https://github.com/pytorch/pytorch/pull/155118.

will close this for now
